platform/src/server/http/http_server.ts

```diff
@@ -1,95 +1,243 @@
-import express from 'express';
-import http from 'http';
-import https from 'https';
-import { Server } from 'net';
-import { promisify } from 'util';
+import { Request, ResponseToolkit, Server, ServerOptions } from 'hapi';
+import { ServerOptions as TLSOptions } from 'https';
 import { readFileSync } from 'fs';
+import { format as formatUrl } from 'url';
 
 import { HttpConfig } from './http_config';
 import { Env } from '../../config';
 import { Logger } from '../../logging';
-import { Router } from './Router';
+import { Router } from './router';
+import { modifyUrl } from '../../lib/utils/url';
 
 export class HttpServer {
-  private readonly app: express.Application;
-  private server?: Server;
+  private _server?: Server;
```

Comment:
Is there a reason we're using `_server` with `private`? I suspect it's to inform people with access to this instance who aren't using TypeScript, but we shouldn't be exposing this instance to non-internal code right? I would prefer that we have a separate public contract from the internal server instance.

Reply:
The were two reasons:

* If we decide to transpile code to JavaScript once again, we'll have to prefix all private variables manually (that was one of the pain point when transpiled new platform to JS), so I thought it'd make sense to keep code in a "close-to-JS" state even if we use TS;

  Same for JSDoc - previously we decided not to mention types in JSDoc for TypeScript, but once I transpiled code to JS all type information disappeared and having that info at least in JSDoc would have helped a lot. So I started to write JSDoc as if it was in JS...

* IIRC Kim and @archanid voted for "_"-prefixed TypeScript private members in the past, so it became informatl new-platform-team-agreement but I can't remember whether it was in slack or in some PR :/

We can easily change team-wide agreement if we want, I'm somewhat concerned about the first point, what do you think?

Reply:
re 1: We're going forward with TypeScript now and I'm buying in fully, but if I need to write a script that rewrites all `private` properties to be `_` prefixed in the future I'm down with that.

re 2: 🤷‍♂️ we have more formally agreed to try and follow `tslint:recommended` rules and they explicitly forbid underscore prefixed variables and properties.

Reply:
> I'm buying in fully, but if I need to write a script that rewrites all private properties to be _ prefixed in the future I'm down with that.

Now we're talking, we such level of confidence in TypeScript, I'm fine with that too :) What about types in JSDoc by the way?

> that preset explicitly forbids underscore prefixed variables and properties.

Oh, interesting, didn't know that.

Reply:
> What about types in JSDoc by the way?

That's a harder one, but since it's not checked by the compiler I don't think we should be putting them in there. Maybe that means we should enable https://palantir.github.io/tslint/rules/no-redundant-jsdoc/ at some point, or maybe that means we should just let JSDoc continue to be a suggestion that may give insight into the authors thought process while also being entirely incorrect at times.

I'm not confident in my ability to automatically create JSDoc comments based on the TS types, but I imagine it's possible to some extent. If we get to a future where we need to rip out TypeScript we should look into that more closely. Otherwise I don't think we're any worse off than we are today if we loose all the type information and don't have type info in the JSDoc comments.

Reply:
> while also being entirely incorrect at times.

I can't agree here, such JSDoc just shouldn't pass review, it's _extremely_ useful to have correct type information even if it's not enforced by compiler/builder/whatever especially with the code base of the Kibana's size and complexity :)

> Otherwise I don't think we're any worse off than we are today if we loose all the type information and don't have type info in the JSDoc comments.

Well it will be worse for the "`original JS` -> `TS` -> `transpiled JS`" case if `original JS` currently uses JSDoc with proper types and we remove them while converting to `TS`.

Okay, it sounds like a separate discussion that deserves its own issue/PR, I'll probably keep adding type information in JSDoc where it's reasonable for the time being and take care of it if we decide to enable `no-redundant-jsdoc`.

Reply:
>> while also being entirely incorrect at times.
> I can't agree here, such JSDoc just shouldn't pass review, 

Yeah, but humans 🤓 I agree though, we should have this discussion with a broader audience

Reply:
Haha, I just resort to idealism sometimes when I need to prove my point, that's true :)